<a href="https://colab.research.google.com/github/David-gutierrez-rodriguez/Simulacion-I/blob/main/Eficiencia_MC_y_MC_crudo.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Se dice que el método 1 es más eficiente que el método 2 si
$$\varepsilon = \frac{var(\theta_1)t_1}{var(\theta_2)t_2} < 1$$

In [2]:
import numpy as np
from random import random
import matplotlib.pyplot as plt
import time

In [3]:
def g(x):
  return np.sqrt(np.atan(x))

Método de Montecarlo

In [1]:
def is_below(x, y):
  c = 0
  for i in range(len(x)):
    if y[i] <= g(x[i]):
      c += 1
  return c

def MC(a, b, N):
  x = np.linspace(a, b, 100)
  y_max = np.max(g(x))
  y_min = np.min(g(x))

  x = [random() * (b - a) + a for _ in range(N)]
  y = [random() * (y_max - y_min) + y_min for _ in range(N)]

  return (b - a) * (y_max - y_min) * is_below(x, y) / N

Método de Monte Carlo crudo

In [4]:
def MC_crudo(a, b, N):
  I = 0
  for i in range(N):
    I += g(random()*(b-a)+a)
  return I/N

Tiempo de ejecución de un método

In [7]:
def tiempo(metodo, n, N):
  t_f = time.time()
  valores = [metodo(0, 1, n) for _ in range(N)]
  t_i = time.time()
  return np.mean(valores), np.var(valores), -t_f + t_i

Realizamos el experimento con $n = 1000$ muestras y lo repetimos $N = 1000$ veces en ambos casos

In [8]:
theta_1, var_theta_1, t_1 = tiempo(MC, 1000, 1000)
theta_2, var_theta_2, t_2 = tiempo(MC_crudo, 1000, 1000)

In [9]:
epsilon = (t_1*var_theta_1)/(t_2*var_theta_2)

print(f"theta_1 = {theta_1}")
print(f"var(theta_1) = {var_theta_1}")
print(f"t_1 = {t_1}\n")
print(f"theta_2 = {theta_2}")
print(f"var(theta_2) = {var_theta_2}")
print(f"t_2 = {t_2}\n")
print(f"epsilon = {epsilon}")

theta_1 = 0.6301064577699854
var(theta_1) = 0.00016783880133447318
t_1 = 3.0657167434692383

theta_2 = 0.6296929787073782
var(theta_2) = 4.2324715782713323e-05
t_2 = 3.3558993339538574

epsilon = 3.6226086488753735


Ya que $\varepsilon$ = 3.62 > 1, entonces el Método de Monte Carlo crudo es más eficiente que el Método de Monte carlo